### PR TITLE
cmd, core/state, eth, tests, trie: improve state reader

### PR DESCRIPTION
**Pull request by rjl493456442**
The state availability is checked when a state reader is requested to be created.
 
- In hash-based database, if the specified root node is not existent in disk, then 
the state reader won't be created and an error will be returned.

- In path-based database, if the specified state layer is not available, then the 
state reader won't be created and an error will be returned.

### CODE REVIEW

File: core/state/statedb_test.go [hunk 1]
```diff
@@ -558,20 +559,6 @@ func TestCopyCommitCopy(t *testing.T) {
 	if val := copyOne.GetCommittedState(addr, skey); val != (common.Hash{}) {
 		t.Fatalf("first copy pre-commit committed storage slot mismatch: have %x, want %x", val, common.Hash{})
 	}
-
-	copyOne.Commit(false)
-	if balance := copyOne.GetBalance(addr); balance.Cmp(big.NewInt(42)) != 0 {
```
Comment (holiman) [inline core/state/statedb_test.go:563]: Why remove this chunk of code?

File: trie/trie_reader.go [hunk 2]
```diff
@@ -46,10 +39,16 @@ type trieReader struct {
...
 
 // newTrieReader initializes the trie reader with the given node reader.
-func newTrieReader(stateRoot, owner common.Hash, db NodeReader) (*trieReader, error) {
-	reader := db.Reader(stateRoot)
-	if reader == nil {
-		return nil, fmt.Errorf("state not found #%x", stateRoot)
+func newTrieReader(stateRoot, owner common.Hash, db *Database) (*trieReader, error) {
+	if stateRoot == (common.Hash{}) || stateRoot == types.EmptyRootHash {
+		if stateRoot == (common.Hash{}) {
+			log.Warn("Initialize trie with empty root")
```
Comment (holiman) [inline trie/trie_reader.go:45]: Please make a more descriptive error message. I don't really understand what this warning is about ...?

File: trie/triedb/hashdb/database.go [hunk 3]
```diff
@@ -621,8 +622,12 @@ func (db *Database) Scheme() string {
 }
 
 // Reader retrieves a node reader belonging to the given state root.
-func (db *Database) Reader(root common.Hash) *reader {
-	return &reader{db: db}
+// An error will be returned if the requested state is not available.
+func (db *Database) Reader(root common.Hash) (*reader, error) {
+	if _, err := db.Node(root); err != nil {
+		return nil, fmt.Errorf("state %x is not available", root)
```
Comment (holiman) [inline trie/triedb/hashdb/database.go:628]: We are dropping `err` on the floor here, why not 
```suggestion
		return nil, fmt.Errorf("state %#x is not available: %v", root, err)
```
Also in general, I prefer to use `%#x` over `%x` , to get the `0x` prefix

File: core/state/statedb_test.go (hunk 1 above)
Comment (rjl493456442) [reply core/state/statedb_test.go:563]: Because once the statedb is committed, then we can't create a copy from it.

The deep reason for this behavior limitation is once the trie(s) are committed, these trie(s) are unusable. We need to re-create statedb to re-initialize all tries with new root upon the updated database.

So yeah, we can't support commit-then-create-a-copy operation.

File: trie/trie_reader.go (hunk 2 above)
Comment (rjl493456442) [reply trie/trie_reader.go:45]: I will change it to `log.Error("Zero state root hash!")`. We suppose to use `types.EmptyRootHash` instead of `common.Hash{}` to represent an empty hash. So just add a warning or error which is not supposed to be printed.

File: core/state/statedb_test.go (hunk 1 above)
Comment (rjl493456442) [reply core/state/statedb_test.go:563]: The state db is still usable after commit is: the state objects are still cached in the state db.
The state db is not usable by copying from a committed one is: the cached objects will be discarded and the copied trie is already non-functional.

I will not encourage people to still use the statedb after commit, it just kinda of work, but very hacky.


> I don't think we should just remove the code which tests non-supported order of operations, we should put it somewhere, so we know what happens if the api is misused this way.

I can add a test to show that statedb copied from the committed one is not functional.

File: core/state/statedb.go [hunk 4]
```diff
@@ -56,8 +56,13 @@ func (n *proofList) Delete(key []byte) error {
 // StateDB structs within the ethereum protocol are used to store anything
 // within the merkle trie. StateDBs take care of caching and storing
 // nested states. It's the general query interface to retrieve:
+//
 // * Contracts
 // * Accounts
+//
+// Once the state is committed, it's not usable anymore. A new state instance
+// must be created with new root and updated database for accessing latest
+// states.
```
Comment (gballet) [inline core/state/statedb.go:65]: that's a big problem for me, as this means calling `OpenTrie`, which becomes an expensive operation with verkle. Is there a way that the trie could be reused?

Comment (rjl493456442) [reply core/state/statedb.go:65]: In our case, we always re-create a new state-db after processing a block, even now.

Re reusing the trie, it will be slightly ugly. Perhaps we can introduce some cache layer above the trie database which caches the decoded nodes. So that reloading trie node can be cheap?

**Comment by holiman** [comment]
For me, it's still a bit unclear what `Once the state is committed, it's not usable anymore` means, in practice. 


For example, `cmd/evm/runner.go` does this: 
```
	if ctx.Bool(DumpFlag.Name) {
		statedb.Commit(true)
		statedb.IntermediateRoot(true)
		fmt.Println(string(statedb.Dump(nil)))
	}
```
Is it "wrong"? Is it relying on undefined behaviour but is ok? Will it misbehave? 


And in  `tests/state_test_util.go`, we return a `statedb` which has already committed: 
```golang
	root, _ := statedb.Commit(config.IsEIP158(block.Number()))
	return snaps, statedb, root, err
}
````
Later on, that `statedb` is used, for intermediate root and rawdumping: 
```golang
			_, s, err := test.Run(st, cfg, false)
			// print state root for evmlab tracing
			if s != nil {
				root := s.IntermediateRoot(false)
..
				if dump && s != nil {
					dump := s.RawDump(nil)
```
And in `cmd/evm/internal/t8ntool/execution.go`, we commit and later ask for logs
```
	root, err := statedb.Commit(chainConfig.IsEIP158(vmContext.BlockNumber))
	if err != nil {
		return nil, nil, NewError(ErrorEVM, fmt.Errorf("could not commit state: %v", err))
	}
	execRs := &ExecutionResult{
...
		LogsHash:    rlpHash(statedb.Logs()),
```
We do some funky commit then commit in `core/chain_makers.go`

```golang
			// Write state changes to db
			root, err := statedb.Commit(config.IsEIP158(b.header.Number))
			if err != nil {
				panic(fmt.Sprintf("state write error: %v", err))
			}
			if err := statedb.Database().TrieDB().Commit(root, false); err != nil {
				panic(fmt.Sprintf("trie write error: %v", err))
			}

```
So, TLDR is that it needs to be more obvious what is allowed vs what is not allowed, and potentially we should even make it explicit: e.g. we could `panic` if the api is misused (or maybe, in production code, print a very stern warning is better than panic)

File: trie/trie_reader.go [hunk 5]
```diff
@@ -46,10 +39,16 @@ type trieReader struct {
...
+func newTrieReader(stateRoot, owner common.Hash, db *Database) (*trieReader, error) {
+	if stateRoot == (common.Hash{}) || stateRoot == types.EmptyRootHash {
+		if stateRoot == (common.Hash{}) {
+			log.Error("Zero state root hash!")
+		}
+		return &trieReader{owner: owner}, nil
+	}
+	reader, err := db.Reader(stateRoot)
+	if err != nil {
+		return nil, &MissingNodeError{Owner: owner, NodeHash: stateRoot, err: err}
```
Comment (holiman) [inline trie/trie_reader.go:51]: Note for reviewers: This is the big functional change, as far as I understand. This change makes it so that a Commit-then-Copy leads to a non-functional statedb, since the trieReader does not actually have a `reader`.

**Comment by rjl493456442** [comment]
> e.g. we could panic if the api is misused (or maybe, in production code, print a very stern warning is better than panic)

I will think about it. I am not a big fan of this approach. But I totally get it. I didn't do it because we need a marker to represent this state-db is committed and following operations should be rejected. It's a big change then.

But now we just assume that users won't use committed state-db, and a part of APIs will still work, and may not work(e.g. a copied state-db from the committed one). It's just too implicit. 

Maybe I need to find a line, that this API is not usable after commit, and these APIs can still work.

**Comment by holiman** [comment]
I'm curious though... Those examples I listed, what is the status on them? Ok or Not?

**Comment by rjl493456442** [comment]
@holiman `And in tests/state_test_util.go, we return a statedb which has already committed:`

The case you mentioned above is actually be fixed by re-initialize a new one here https://github.com/rjl493456442/go-ethereum/blob/fix-state-commit-2/tests/state_test_util.go#L208

**Comment by rjl493456442** [comment]
@holiman nope, I will deploy in on 08 by running snap sync
